project-chip/connectedhomeip
# Handle fabric removal in the Door Lock Cluster and fix GetUser/GetCredentialStatus commands

#### Problem
* Door lock cluster did not handle the fabric removal properly.
* Fixes #14919.

#### Change overview
* Now door lock cluster fixes the fabric removal according to spec: 
  * If the fabric was deleted, the creator/modifier fabric index in user/credential records is set to 0 (both supported in SDK and Linux App)
  * If the credential was created outside Matter IM the creator/modifier fabric index in user/credential records is set to null (only SDK support as there's no way to create users/credentials outside Matter IM for now in the Linux App).
* GetUserResponse/GetCredentialStatusResponse status command now return next *occupied* index in the payload.

#### Testing
* Ran DL_UsersCredentials, DL_LockUnlock and DL_Schedules YAML tests.
* To test the fabric removal I've used the following procedure:
  1. Run the Linux `chip-lock-app`
  2. Pair with the app using the `chip-tool`:
     ```
     ? ./chip-tool pairing onnetwork 1 20202021
     ```
  3. Open commissioning window again so we can pair with another fabric:
     ```
     ? ./chip-tool pairing open-commissioning-window 1 0 200 1000 3840
     ```
  4. Pair with different fabric:
     ```
     ? ./chip-tool pairing onnetwork 1 20202021 --commissioner-name beta
     ```
  5. Create new credentials and users from the first fabric: 
     ```
     ? ./chip-tool doorlock set-credential 0 '{ "credentialType" : 1 , "credentialIndex" : 1 }' 123456 null null null 1 1 --timedInteractionTimeoutMs 1000
     ```
  6. Create new credentials and users from the second fabric:
     ```
     ? ./chip-tool doorlock set-credential 0 '{ "credentialType" : 2 , "credentialIndex" : 2 }' 12345678901213 null null null 1 1 --timedInteractionTimeoutMs 1000 --commissioner-name beta
     ```
  7. Delete the second fabric:
     ```
     ? ./chip-tool operationalcredentials remove-fabric 2 1 0 --commissioner-name beta
     ```
  8. Make sure the fabric index in the users/credentials created by the first fabric is the same:
     ```
     ? chip-tool doorlock get-user 1 1 1
     ? chip-tool doorlock get-credential-status '{ "credentialType" : 1 , "credentialIndex" : 1 }' 1 1
     ```
  9. Make sure the fabric index in the users/credentials created by the second fabric is set to `0`:
     ```
     ? chip-tool doorlock get-user 2 1 1
     ? chip-tool doorlock get-credential-status '{ "credentialType" : 2 , "credentialIndex" : 2 }' 1 1
     ```
